### PR TITLE
[Android] Fix audio engine initialisation after sleep

### PR DESCRIPTION
**Pull request by hugbug**
## Description

Audio engine is reinitialised when event ACTION_HDMI_AUDIO_PLUG is received.

## Motivation and context

When a device goes to sleep the audio engine is suspended. On wake-up the audio engine is initialised again (resumed). Wake-up processing is performed on event ACTION_SCREEN_ON. There is another event (ACTION_HDMI_AUDIO_PLUG), which system also sends on wake-up. At least on my system the HDMI handshake isn't completed when the first event arrives. As a result the audio engine isn't properly initialised. This causes all kind of issues such as:

- Kodi IEC packer isn't listed in audio device list
- Audio passthrough isn't working anymore
- Audio is played as PCM stereo only, etc.

The problem is reported by various users on forum since 2018 - [Audio Passthrough gone after Shield Sleep](https://forum.kodi.tv/showthread.php?tid=333546). It is not Shield specific though. I have it with Fire TV 4K Max.

When the event ACTION_HDMI_AUDIO_PLUG is received, the initialisation of audio engine can be performed properly.

I considered moving audio engine initialisation from ACTION_SCREEN_ON to ACTION_HDMI_AUDIO_PLUG but decided against it. This would make the code more complex and requires changes in Power Manager. I'm also not sure if the second event is presented on all devices. I guess Android phones will not get it, as opposed to ACTION_SCREEN_ON, which probably every device receives.

What the fix does instead: on receiving ACTION_HDMI_AUDIO_PLUG event the audio engine is reinitialised using Suspend/Resume. This approach seems to be more robust and will hopefully not break anything.

## How has this been tested?

Fire TV Stick 4K Max 2nd gen, connected to AVR Denon AVC-X4800H, connected to TV LG C1. Fire TV is configured to stop outputting signal on sleep. Fire TV and Kodi are configured to output all audio formats as passthrough, using Kodi IEC packer.

#### Test 1

1. Start Kodi
2. Make sure audio passthrough is configured in Kodi settings, using Kodi IEC packer.
3. Playback a video with multichannel audio such as Dolby Digital or DTS, check on AVR that the incoming signal is correct
4. Stop video playback
5. On Fire TV remote control click the gear-button and choose "Sleep" in the menu 
6. Wait a few seconds for sleep mode to activate
7. AVR and TV remain still on in this test
8. Wake up Fire TV by clicking select-button on the remote
9. Kodi should be active
10. Try to playback the same video
11. Check incoming signal on AVR.
12. Go to Audio settings in Kodi. Check that Kodi IEC packer is still selected as audio passthrough device.

#### Test 2

This test is similar to Test 1 but on step 5 instead of putting only Fire TV to sleep we switch off all devices using on/off button on Fire TV remote. If Fire TV is properly configured the AVR and TV will both switch off. On step 8 we use the same on/off button on remote to wake up Fire TV and switch on AVR and TV.

##### Before fix
On step 11 AVR shows input signal as "PCM Stereo". On step 12 "Kodi IEC packer" disappears from the list of audio devices. Android IEC packer is shown instead (it doesn't work properly on Fire TV).

##### After fix
On step 11 AVR shows the same input signal as on step 3. On step 12 "Kodi IEC packer" is in the list of audio devices and it is selected as output device. Audio passthrough works correctly.

## What is the effect on users?

Users who were having problem with passthrough not working after sleep should have the issue fixed. In particular Shield and Fire TV devices (especially if connected to AVR) are known to be affected.

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/platform/android/activity/XBMCApp.cpp [hunk 1]
```diff
@@ -1333,6 +1333,12 @@ void CXBMCApp::onReceive(CJNIIntent intent)
       if (winSystem && dynamic_cast<CWinSystemAndroid*>(winSystem))
         dynamic_cast<CWinSystemAndroid*>(winSystem)->SetHdmiState(hdmiPlugged);
     }
+    if (hdmiPlugged)
+    {
+      CLog::Log(LOGDEBUG, "{}: Reset audio engine", __FUNCTION__);
+      CServiceBroker::GetActiveAE()->Suspend();
+      CServiceBroker::GetActiveAE()->Resume();
```
Comment (fritsch) [inline xbmc/platform/android/activity/XBMCApp.cpp:1340]: CServiceBroker::GetActiveAE()->DeviceChange(); should also work and semantically do the right thing. Suspend / Resume works by chance as Resume causes an internal reenumeration.

Bonus-Points:
As we basically only want to rescan when reinitialize AE when really something changed, you could compare the information you get with the HDMI_PLUG callback to see if something new pops up and only then issue the CServiceBroker::GetActiveAE()->DeviceChange(); call.

Comment (fritsch) [reply xbmc/platform/android/activity/XBMCApp.cpp:1340]: That being said: Ignore the Bonus Points for now ;-). We can see if it's needed when reports on the forum popup. Only change the Suspend / Resume please to DeviceChange(); thanks!

File: xbmc/platform/android/activity/XBMCApp.cpp [hunk 2]
```diff
@@ -1333,6 +1333,12 @@ void CXBMCApp::onReceive(CJNIIntent intent)
       if (winSystem && dynamic_cast<CWinSystemAndroid*>(winSystem))
         dynamic_cast<CWinSystemAndroid*>(winSystem)->SetHdmiState(hdmiPlugged);
     }
+    if (hdmiPlugged)
+    {
+      CLog::Log(LOGDEBUG, "{}: Reset audio engine", __FUNCTION__);
```
Comment (thexai) [inline xbmc/platform/android/activity/XBMCApp.cpp:1338]:
```suggestion
      CLog::Log(LOGDEBUG, "CXBMCApp::{}: Reset audio engine", __FUNCTION__);
```

**Comment by hugbug** [comment]
Thanks for suggestions. I've applied the changes.

**Review by fritsch — APPROVED** [review]
Works for me.

**Comment by hugbug** [comment]
As you mentioned on forum the resolution or refresh rate changes on begin of playback may fire ACTION_HDMI_AUDIO_PLUG events. That indeed was happening causing a slight delay (300-500ms) at the begin of playback every time, due to AE initialisation. In order to avoid this I've added a check that ensures the audio engine initialisation is performed only when ACTION_HDMI_AUDIO_PLUG is received after ACTION_SCREEN_ON. That check is much simpler than comparing device capabilities. 

The new fix works just fine in my tests.

Please let me know if you want the two commit squashed. Thanks again for the review.

**Comment by fritsch** [comment]
On Refreshrate change kodi has a different infrastructure with e.g. OnLostDevice where we also use the hdmi plug event on android. Getting this perfectly right is not an easy task. Doing double / triple enumerations in that case is not correct, yes.

The best solution would be to try to use:
https://developer.android.com/reference/android/media/AudioManager#EXTRA_ENCODINGS

If cached list does not contain IEC and IEC gets available -> device change, else not and only issue DeviceChange when the entries have changed.

Alternatively:
For  https://developer.android.com/reference/android/content/Intent#ACTION_SCREEN_ON ... maybe it might be even "more wise" to just issue DeviceChange when kodi becomes interactive, e.g. not using HDMI_PLUG but simply ACTION_SCREEN_ON?

**Comment by fritsch** [comment]
Just thinking loud above. Your change to combine it with ScreenOn is still okay for me.

**Review by thexai — APPROVED** [review]
Tested previous version in Shield and not caused issues. Then this less aggressive version, if still fix the sleep issue, seems even better.

**Comment by fritsch** [comment]
Here we go. Thanks for your contribution - especially a nice usability fix at the end of this year. All the best - if you find other places you want to improve / help out / please keep them coming.

**Pull request merged**